File: java/01-bowling/src/main/java/bowling/Game.java
package bowling;

public class Game {
    private int[] rolls = new int[21];
    private int headIndex;

    public void roll(int numPins) {
        rolls[headIndex] = numPins;
        headIndex++;
    }

    public int calculateScore() {
        int sum = 0;
        for (int frame = 0, frameIndex = 0; frame < 10; frame++) {
            if (isStrike(frameIndex)) {
                sum += scoreStrike(frameIndex);
                frameIndex++;
            } else if (isSpare(frameIndex)) {
                sum += scoreSpare(frameIndex);
                frameIndex += 2;
            } else {
                sum += sumOfRollsInFrame(frameIndex);
                frameIndex += 2;
            }
        }
        return sum;
    }

    private int sumOfRollsInFrame(int frameIndex) {
        return rolls[frameIndex] + rolls[frameIndex + 1];
    }

    private int scoreSpare(int frameIndex) {
        return 10 + rolls[frameIndex + 2];
    }

    private int scoreStrike(int frameIndex) {
        return 10 + rolls[frameIndex + 1] + rolls[frameIndex + 2];
    }

    private boolean isStrike(int frameIndex) {
        return rolls[frameIndex] == 10;
    }

    private boolean isSpare(int frameIndex) {
        return sumOfRollsInFrame(frameIndex) == 10;
    }
}

File: kotlin/01-bowling/src/main/kotlin/bowling/Game.kt
package bowling

class Game {

    private var rolls: IntArray = IntArray(21)
    private var currentRoll: Int = 0

    val score: Int
        get() {
            var frameIndex = 0
            var score = 0
            for (frame in 1..10) {
                when {
                    isStrike(frameIndex) -> {
                        score += scoreStrike(frameIndex)
                        frameIndex++
                    }
                    isSpare(frameIndex) -> {
                        score += scoreSpare(frameIndex)
                        frameIndex += 2
                    }
                    else -> {
                        score += sumOfRollsInFrame(frameIndex)
                        frameIndex += 2
                    }
                }
            }
            return score
        }

    private fun isStrike(frameIndex: Int) = rolls[frameIndex] == 10

    private fun scoreStrike(frameIndex: Int) = 10 + rolls[frameIndex + 1] + rolls[frameIndex + 2]

    private fun isSpare(frameIndex: Int) = sumOfRollsInFrame(frameIndex) == 10

    private fun scoreSpare(frameIndex: Int) = 10 + rolls[frameIndex + 2]

    private fun sumOfRollsInFrame(frameIndex: Int) = rolls[frameIndex] + rolls[frameIndex + 1]

    fun roll(numPins: Int) {
        rolls[currentRoll] = numPins
        currentRoll++
    }
}
